VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol

```diff
@@ -0,0 +1,1098 @@
+            if (block.timestamp - earlistTime < 24 hours) {
+                revert TimeNotElapsed();
+            }
+        }
+        // Check 1 hour condition for same node
+        for (uint i; i < logs._issueList.length;) {
+            if (logs._issueList[i].nodeDID == nodeAddress && 
+                (block.timestamp - logs._issueList[i].time) < _slotInfo.SAME_NODE_LOG_DURATION) {
+                revert InvalidSameNodeTime();
+            }
+            unchecked { ++i; }
+        }
+
+        // Add or update
+        if (logs._issueList.length < 4) {
+            logs._issueList.push(IssueInformation(nodeAddress, reasonCode, block.timestamp));
+        } else {
+            uint index = logs.index;
+            logs._issueList[index].nodeDID = nodeAddress;
+            logs._issueList[index].reasonCode = reasonCode;
+            logs._issueList[index].time = block.timestamp;
+            ++index;
+            logs.index = index % 4;
+        }
+
+        // Transfer fees to this contract
+        IERC20Upgradeable(vdaTokenAddress).transferFrom(msg.sender, address(this), _slotInfo.NODE_ISSUE_FEE);
+
+        _slotInfo.totalIssueFee = _slotInfo.totalIssueFee + _slotInfo.NODE_ISSUE_FEE;
+
+        uint val;
+        EnumerableMapUpgradeable.AddressToUintMap storage didReasonLogAmount = _loggedTokenAmount[nodeAddress][reasonCode];
+        if (didReasonLogAmount.contains(didAddress)) {
+            val = didReasonLogAmount.get(didAddress);
+        }
+
+        didReasonLogAmount.set(didAddress, val + _slotInfo.NODE_ISSUE_FEE);
+        _issueTotalAmount[nodeAddress][reasonCode] = _issueTotalAmount[nodeAddress][reasonCode] + _slotInfo.NODE_ISSUE_FEE;
+
+        emit LoggedNodeIssue(didAddress, nodeAddress, reasonCode);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function slash(
+        address nodeDID,
+        uint reasonCode,
+        uint amount,
+        string calldata moreInfoUrl
+    ) external payable onlyOwner {
+
+        if (amount == 0 || amount > _stakedTokenAmount[nodeDID]) {
+            revert InvalidAmount();
+        }
+
+        uint issueAmount = _issueTotalAmount[nodeDID][reasonCode];
+        if (issueAmount == 0) {
+            revert InvalidReasonCode();
+        }
+
+        _stakedTokenAmount[nodeDID] = _stakedTokenAmount[nodeDID] - amount;
+        
+        EnumerableMapUpgradeable.AddressToUintMap storage logInfo = _loggedTokenAmount[nodeDID][reasonCode];
+        uint loggerCount = logInfo.length();
+        uint distributeTotalAmount;
+
+        for (uint i; i < loggerCount;) {
+            (address didAddress, uint loggerStaked) = logInfo.at(i);
+            uint distAmount = amount * loggerStaked / issueAmount;
+            distributeTotalAmount += distAmount;
+
+            _stakedTokenAmount[didAddress] = _stakedTokenAmount[didAddress] + distAmount;
+
+            unchecked { ++i; }
+        }
+
+        emit Slash(nodeDID, reasonCode, distributeTotalAmount, moreInfoUrl);
```

Comment:
Also emit how many `didAddress`'s received tokens.

Should we also emit a `loggedIssueDeposit()` event to each `didAddress`? That way a `didAddress` can monitor for events where they've been rewarded for logging an issue?

Reply:
We can emit how many DID addresses received tokens.

By the way, it's not conveninen to emit per each address. Emitting an event also requires gas fee.
I think, the users can monitor the logged DID & Reason code to know that they're rewarded. After that, they can know their excess token amounts.

Reply:
Ok, sounds good.